cumulus_library/studies/core/core_templates/core_utils.jinja

```diff
@@ -1,38 +1,138 @@
+{#- A collection of macros for constructing common safe column selects
+
+    For layout, is is generally assumed that these are being invoked from within
+    a WITH TABLE AS(...) statement, so they will always assume two levels of
+    indentation.
+-#}
+
 {%- macro comma_delineate(loop) -%}
 {%- if not loop.last -%}
 , 
 {%- endif -%}
 {%- endmacro -%}
 
--- gets a list of columns from a known present source
+{#- gets a list of columns from a known present source -#}
 {% macro basic_cols(table, alias, targets) %}
         {%- for col in targets %}
         {{ alias }}.{{ col.lower() }}
         {{- comma_delineate(loop) }}
         {%- endfor %}
 {%- endmacro %}
 
+{#- gets a list of columns from a potentially null source.
+
+targets is an array expecting a data type of the following:
+    string: refers to a vanilla SQL primitive column
+    tuple(col, object, col_alias): refers to a one level nested SQL object
+    tuple(col, object, , child, col_alias): refers to a two level nested SQL object
 
--- gets a column, if present, from a basic SQL primitive col, or a null otherwise
-{% macro basic_nullable_cols(table, alias, targets, schema) %}
+-#}
+{% macro nullable_cols(table, alias, targets, schema) %}
         {%- for col in targets %}
-        {%- if schema[table][col] %}
+        {#- depth one nested column-#}
+        {%- if col is not string and col|length ==3%}
+        {%- if schema[table][col[0]][col[1]] %}
+        {{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }} AS {{col[2]}}
+        {%- else %}
+        NULL AS {{col[2]}}
+        {%- endif %}
+        {#- depth two nested column -#}
+        {%- elif col is not string and col|length ==4%}
+        {%- if schema[table][col[0]][col[2]] %}
+        {{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }}.{{ col[2].lower() }} AS {{col[3]}}
+        {%- else %}
+        NULL AS {{col[3]}}
+        {%- endif %}
+        {#- SQL primitive column column-#}
+        {%- elif schema[table][col] %}
         {{ alias }}.{{ col.lower() }}
         {%- else %}
         NULL AS {{ col.lower() }}
+        {%- endif %}        
+        {{- comma_delineate(loop) }}
+        {%- endfor %}
+{%- endmacro %}
+
+{#- creates a date from a from an ISO formatted date str 
+targets is an array expecting a data type of the following:
+    string: refers to a vanilla SQL primitive column
+    tuple(col, object, col_alias): refers to a one level nested SQL object
+    tuple(col, object, , child, col_alias): refers to a two level nested SQL obj
+-#}
+{% macro date_cols_from_str(table, alias, targets, schema) %}
+        {%- for col in targets %}
+        {#- depth one nested column-#}
+        {%- if col is not string and col|length ==3%}
```

Comment:
super nit: why do you never leave a space after the number on these?

Reply:
because sqlfluff will complain and force me to fix it/fix it for me. 

i'm glad we don't do pair programming often, because the amount of stuff like this i do without thinking about it, throwing it off to a linter to handle later, would make you %100 insane. like mixing single quotes and double quotes in the same line! i'm a monster, but at least i know enough to set traps for myself.

Reply:
:smile:

Are you saying that sqlfluff will fix this for you so you don't worry about it (in which case, this is going to be fixed later as you work on linting errors?), or are you saying that sqlfluff *wants* there to be no space?

Reply:
the former.